### PR TITLE
refactor: add Java constrainer

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds Java constrainer.

**Related issue(s)**
Blocked by #683

### CODE REVIEW

**Review by magicmatatjahu — COMMENTED** [review]
Two comments, I think they are important :)

File: src/generators/java/JavaConstrainer.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+  const constrainedModel = new ConstrainedTupleModel(constrainedName, metaModel.originalInput, '', constrainedTupleModels);
+  if (typeMapping.Tuple !== undefined) {
+    constrainedModel.type = typeMapping.Tuple(constrainedModel);
+  } else {
+    //Because Java have no notion of tuples (and no custom implementation), we have to render it as a list of any value.
+    const tupleType = 'Object';
+    if (options.collectionType && options.collectionType === 'List') {
+      constrainedModel.type = `List<${tupleType}>`;
+    }
+    constrainedModel.type = `${tupleType}[]`;
```
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/generators/java/JavaConstrainer.ts:147]: What do you mean exactly? 🙂

File: src/generators/java/JavaConstrainer.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+function constrainArrayModel(constrainedName: string, metaModel: ArrayModel, typeMapping: TypeMapping, constrainRules: JavaConstraints, options: JavaOptions): ConstrainedArrayModel {
+  const constrainedValueModel = constrainMetaModel(metaModel.valueModel, typeMapping, constrainRules, options);
+  const constrainedModel = new ConstrainedArrayModel(constrainedName, metaModel.originalInput, '', constrainedValueModel);
+  if (typeMapping.Array !== undefined) {
+    constrainedModel.type = typeMapping.Array(constrainedModel);
+  } else {
+    if (options.collectionType && options.collectionType === 'List') {
+      constrainedModel.type = `List<${constrainedValueModel.type}>`;
+    }
+    constrainedModel.type = `${constrainedValueModel.type}[]`;
```
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/generators/java/JavaConstrainer.ts:160]: What do you mean exactly? 🙂

File: src/generators/java/JavaConstrainer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/generators/java/JavaConstrainer.ts:147]: OH! You right! Changing 😆 

Good catch!

**Review by magicmatatjahu — APPROVED** [review]

**Comment by jonaslagoni** [comment]
Thanks for the quick review @magicmatatjahu 🙇

**Comment by magicmatatjahu** [comment]
@jonaslagoni Yeah quick 😆

**Pull request merged**